pkg/eventfilter/subscriptionsapi/all_filter.go

```diff
@@ -18,34 +18,88 @@ package subscriptionsapi
 
 import (
 	"context"
+	"sync"
 
 	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"go.uber.org/atomic"
 	"go.uber.org/zap"
 	"knative.dev/pkg/logging"
 
 	"knative.dev/eventing/pkg/eventfilter"
 )
 
-type allFilter []eventfilter.Filter
+type filterCount struct {
+	filter eventfilter.Filter
+	count  atomic.Uint64
+}
+
+type allFilter struct {
+	filters []filterCount
+	rwMutex sync.RWMutex
+	c       chan int
+	d       chan bool
+}
 
 // NewAllFilter returns an event filter which passes if all the contained filters pass
 func NewAllFilter(filters ...eventfilter.Filter) eventfilter.Filter {
-	return append(allFilter{}, filters...)
+	filterCounts := make([]filterCount, len(filters))
+	for i, filter := range filters {
+		filterCounts[i] = filterCount{
+			count:  *atomic.NewUint64(uint64(0)),
+			filter: filter,
+		}
+	}
+	f := &allFilter{
+		filters: filterCounts,
+		c:       make(chan int),
+		d:       make(chan bool),
+	}
+	go f.optimizeLoop()
```

Comment:
when does this go routing terminate?

Reply:
Nevermind, I see the `more` logic

Reply:
I'll add a comment that this gets terminated when `Cleanup` is called